// snarkyjs/dist/node/provable/transaction-leaves.js
import { Field, Bool } from '../lib/core.js';
import { UInt32, UInt64, Sign } from '../lib/int.js';
import { PublicKey } from '../lib/signature.js';
import { derivedLeafTypes } from './derived-leaves.js';
import { createEvents, dataAsHash } from '../lib/events.js';
import { Poseidon } from '../lib/hash.js';
import { provable } from '../lib/circuit_value.js';
export { PublicKey, Field, Bool, AuthRequired, AuthorizationKind, UInt64, UInt32, Sign, TokenId, };
export { Events, SequenceEvents, StringWithHash, TokenSymbol, SequenceState };
const { TokenId, TokenSymbol, AuthRequired, AuthorizationKind } = derivedLeafTypes({ Field, Bool });
const { Events, SequenceEvents } = createEvents({ Field, Poseidon });
const SequenceState = {
    ...provable(Field),
    emptyValue: SequenceEvents.emptySequenceState,
};
const StringWithHash = dataAsHash({
    emptyValue() {
        return {
            data: '',
            hash: Field('22930868938364086394602058221028773520482901241511717002947639863679740444066'),
        };
    },
    toJSON(data) {
        return data;
    },
    fromJSON(json) {
        let data = json;
        // TODO compute hash
        throw Error('unimplemented');
    },
});
