// snarkyjs/dist/node/provable/gen/transaction.js
// @generated this file is auto-generated - don't edit it directly
import { PublicKey, UInt64, UInt32, TokenId, Field, Bool, AuthRequired, TokenSymbol, Sign, AuthorizationKind, StringWithHash, Events, SequenceEvents, SequenceState, } from '../transaction-leaves.js';
import { ProvableFromLayout, } from '../../provable/from-layout.js';
import * as Json from './transaction-json.js';
import { jsLayout } from './js-layout.js';
export { customTypes, ZkappCommand, AccountUpdate };
export { Json };
export * from '../transaction-leaves.js';
export { provableFromLayout, toJSONEssential };
const TypeMap = {
    PublicKey,
    UInt64,
    UInt32,
    TokenId,
    Field,
    Bool,
    AuthRequired,
    Sign,
    AuthorizationKind,
};
let customTypes = {
    StringWithHash,
    TokenSymbol,
    Events,
    SequenceEvents,
    SequenceState,
};
let { provableFromLayout, toJSONEssential } = ProvableFromLayout(TypeMap, customTypes);
let ZkappCommand = provableFromLayout(jsLayout.ZkappCommand);
let AccountUpdate = provableFromLayout(jsLayout.AccountUpdate);
